### PR TITLE
Redesign Differential

### PR DESCRIPTION
**Pull request by HarrisonGrodin**
Redesign representation of `Differential`.

### CODE REVIEW

**Comment by HarrisonGrodin** [comment]
Current `operators.jl -> differentials.jl` diff, since it doesn't show due to file rename:

https://www.diffchecker.com/ch33REKp

**Comment by ChrisRackauckas** [comment]
What's the advantage of this over having a derivative function defined and D(x) being an operation? Or is this just a first step in that direction?

File: src/differentials.jl [hunk 1]
```diff
@@ -0,0 +1,92 @@
+struct Differential <: Expression
+    x::Expression
```
[the remark this answers is not in the document]
Comment (HarrisonGrodin) [reply src/differentials.jl]: Right. I'm currently attempting to refactor the IR such that the types are a bit more correct (which in turn should make a lot of the code cleaner). Once everything seems correct and fully featured, I can focus more on optimizations.

**Comment by HarrisonGrodin** [comment]
This was going to be a step in that direction, with `D` being treated as a higher-order function (once that was supported in IR). Now that you mention it, though, it seems like I could change the subtyping to `<: Function` and get this immediately.

File: src/differentials.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (HarrisonGrodin) [reply src/differentials.jl]: Actually, it seems like we aren't getting small union optimizations with `Union{Variable, Operation}` either:
```julia
julia> Base.isbitsunion(Union{Variable,Operation})
false
```

File: src/differentials.jl [hunk 2]
```diff
@@ -5,7 +5,7 @@ end
 Differential(x) = Differential(x,1)
 
 Base.show(io::IO, D::Differential) = print(io,"($(D.x),$(D.order))")
-Base.Expr(D::Differential) = :($(Symbol("D_$(D.x.name)_$(D.order)")))
+Base.Expr(D::Differential) = D
```
Comment (ChrisRackauckas) [inline src/differentials.jl]: this might be problematic. This could be a good place to use a unicode symbol.

**Comment by ChrisRackauckas** [comment]
This looks like a great first step.

**Pull request merged**